components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java

```diff
@@ -1434,21 +1435,47 @@ public static int getDefaultItemsPerPage() {
     public static boolean isGroupsVsRolesSeparationImprovementsEnabled() {
 
         try {
-            if (AdminServicesUtil.getUserRealm() == null) {
+            UserRealm userRealm = AdminServicesUtil.getUserRealm();
+            if (userRealm == null) {
                 log.warn("Unable to find the user realm, thus GroupAndRoleSeparationEnabled is set as FALSE.");
                 return Boolean.FALSE;
             }
-            diagnosticLog.info("'GroupAndRoleSeparationEnabled' property is set to: " +
-                    UserCoreUtil.isGroupsVsRolesSeparationImprovementsEnabled(AdminServicesUtil.getUserRealm()
-                    .getRealmConfiguration()));
-            return UserCoreUtil.isGroupsVsRolesSeparationImprovementsEnabled(AdminServicesUtil.getUserRealm()
-                    .getRealmConfiguration());
-        } catch (org.wso2.carbon.user.core.UserStoreException | CarbonException e) {
+            return getGroupsVsRolesSeparationImprovementsEnabled(userRealm.getRealmConfiguration());
+
+        } catch (CarbonException | UserStoreException e) {
             log.warn("Property value parsing error: GroupAndRoleSeparationEnabled, thus considered as FALSE");
             return Boolean.FALSE;
         }
     }
 
+    /**
+     * Get the groups vs roles separation config from groupsVsRolesSeparationImprovementsEnabled variable.
+     *
+     * @param realmConfig realmConfiguration.
+     * @return Where groups vs separation enabled or not.
+     */
+    private static boolean getGroupsVsRolesSeparationImprovementsEnabled(RealmConfiguration realmConfig) {
+
+        Boolean isEnabled = groupsVsRolesSeparationImprovementsEnabled.get(realmConfig.getTenantId());
+        if (isEnabled == null) {
+            isEnabled = setGroupsVsRolesSeparationImprovementsEnabled(realmConfig);
+        }
+        return isEnabled;
+    }
+
+    /**
+     * set the groups vs roles separation config to groupsVsRolesSeparationImprovementsEnabled variable.
+     *
+     * @param realmConfig realmConfiguration.
+     * @return Where groups vs separation enabled or not.
+     */
+    private static boolean setGroupsVsRolesSeparationImprovementsEnabled(RealmConfiguration realmConfig) {
```

Comment:
Usually, we use void as the return type for setters.

Reply:
Addressed with commit ec155964c41b15dfd11075b4bcd1bf1c211cf1f3